src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs

```diff
@@ -0,0 +1,226 @@
+            }
+            _targetFrameworks = targetFrameworks;
+
+            if (metadataProvider == null)
+            {
+                throw new ArgumentNullException(nameof(metadataProvider));
+            }
+            _metadataProvider = metadataProvider;
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+            _telemetryService = telemetryService;
+            _logger = logger;
+
+            try
+            {
+                // Get NuGet package recommender service
+                NuGetRecommender = Package.GetGlobalService(typeof(Recommender.SVsNuGetRecommenderService)) as Recommender.IVsNuGetPackageRecommender;
+            }
+            catch (Exception ex)
+            {
+                Debug.WriteLine(ex);
+            }
+        }
+
+        private class RecommendSearchToken : ContinuationToken
+        {
+            public string SearchString { get; set; }
+            public SearchFilter SearchFilter { get; set; }
+        }
+
+        public Task<SearchResult<IPackageSearchMetadata>> SearchAsync(string searchText, SearchFilter searchFilter, CancellationToken cancellationToken)
+        {
+            var searchToken = new RecommendSearchToken
+            {
+                SearchString = searchText,
+                SearchFilter = searchFilter,
+            };
+
+            return RecommendPackagesAsync(searchToken, cancellationToken);
+        }
+
+        public async Task<SearchResult<IPackageSearchMetadata>> RecommendPackagesAsync(ContinuationToken continuationToken, CancellationToken cancellationToken)
+        {
+            var searchToken = continuationToken as RecommendSearchToken;
+            if (searchToken == null)
+            {
+                throw new InvalidOperationException("Invalid token");
+            }
+            // don't make recommendations if the user entered a search string
+            if (searchToken.SearchString != "")
+            {
+                return SearchResult.Empty<IPackageSearchMetadata>();
+            }
+
+            TelemetryServiceUtility.StartOrResumeTimer();
+
+            List<string> recommendIds = new List<string>();
+            try
+            {
+                if (NuGetRecommender != null)
+                {
+                    // get lists of only the package ids to send to the recommender
+                    List<string> topPackages = _installedPackages.Select(item => item.Id.ToLowerInvariant()).ToList();
+                    List<string> depPackages = _dependentPackages.Select(item => item.Id.ToLowerInvariant()).ToList();
+                    // call the recommender to get package recommendations
+                    recommendIds = await NuGetRecommender.GetRecommendedPackagIdsAsync(_targetFrameworks, topPackages, depPackages, cancellationToken);
+                }
+            }
+            catch (Exception ex)
+            {
+                Debug.WriteLine(ex);
```

Comment:
I'd avoid debugs like this. 
If important ensure that they are communicated in an actionable manner. 
Ideally the recommender doesn't throw. 

Is throwing part of the contract?

Reply:
Removed